### PR TITLE
[#5068]  feat(core): support GCS token provider

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
support GCS token provider

### Why are the changes needed?

Fix: #5068 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?

add IT and run with real google acount

### CODE REVIEW

**Comment by FANNG1** [comment]
@jerryshao please help to review when you have time, thanks

**Comment by jerryshao** [comment]
You'd better ping @jerqi and @yuqi1129 review first.

File: iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import org.apache.gravitino.catalog.lakehouse.iceberg.IcebergConstants;
+import org.apache.gravitino.credential.CredentialConstants;
+import org.apache.gravitino.iceberg.common.IcebergConfig;
+import org.apache.gravitino.integration.test.util.DownloaderUtils;
+import org.apache.gravitino.integration.test.util.ITUtils;
+import org.apache.gravitino.storage.GCSProperties;
+import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
+
+// You should export GRAVITINO_GCS_BUCKET and GOOGLE_APPLICATION_CREDENTIALS to run the test
+@EnabledIfEnvironmentVariable(named = "GRAVITINO_TEST_CLOUD_IT", matches = "true")
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java:37]: `GRAVITINO_TEST_CLOUD_IT` is used to enable this test, `GRAVITINO_GCS_BUCKET` is used to set the name of GCS bucket, `GOOGLE_APPLICATION_CREDENTIALS` is used to set the GCS credentials.

File: core/src/main/java/org/apache/gravitino/credential/config/GCSCredentialConfig.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.gravitino.config.ConfigBuilder;
+import org.apache.gravitino.config.ConfigConstants;
+import org.apache.gravitino.config.ConfigEntry;
+import org.apache.gravitino.storage.GCSProperties;
+
+public class GCSCredentialConfig extends Config {
+  public static final ConfigEntry<String> GCS_CREDENTIAL_FILE_PATH =
+      new ConfigBuilder(GCSProperties.GRAVITINO_GCS_CREDENTIAL_FILE_PATH)
+          .doc("The path of GCS credential file")
+          .version(ConfigConstants.VERSION_0_7_0)
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/credential/config/GCSCredentialConfig.java:34]: seems not the scope of this PR

File: iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        IcebergConfig.ICEBERG_CONFIG_PREFIX + GCSProperties.GRAVITINO_GCS_CREDENTIAL_FILE_PATH,
+        gcsCredentialPath);
+    configMap.put(
+        IcebergConfig.ICEBERG_CONFIG_PREFIX + IcebergConstants.IO_IMPL,
+        "org.apache.iceberg.gcp.gcs.GCSFileIO");
+    configMap.put(IcebergConfig.ICEBERG_CONFIG_PREFIX + IcebergConstants.WAREHOUSE, gcsWarehouse);
+    return configMap;
+  }
+
+  private void copyGCSBundleJar() {
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java:88]: ok, I'll check how to resue existing code.

[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java:87]: updated

**Comment by FANNG1** [comment]
@yuqi1129 , comments are addressed, please help to review again

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  public void close() {}
+
+  @Override
+  public String credentialType() {
+    return CredentialConstants.GCS_TOKEN_CREDENTIAL_PROVIDER_TYPE;
+  }
+
+  @Override
+  public Credential getCredential(CredentialContext context) {
+    if (!(context instanceof PathBasedCredentialContext)) {
```
Comment (jerqi) [inline bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java:75]: return null or throw exception?

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                      bucketName, resourcePath));
+              // add list privilege
+              resourceExpressions.add(
+                  String.format(
+                      "api.getAttribute('storage.googleapis.com/objectListPrefix', '').startsWith('%s')",
+                      resourcePath));
+              if (writeLocations.contains(location)) {
+                writeBuckets.add(bucketName);
+                resourceExpressions =
+                    writeExpressions.computeIfAbsent(bucketName, key -> new ArrayList<>());
```
Comment (jerqi) [inline bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java:135]: addd write privilege?

**Comment by jerqi** [comment]
Will the Gravitino fileset reuse this token provider? If so, credential config file isn't a good choice, because if we have different fileset catalog, they maybe have different different GCS credential files.

**Comment by jerqi** [comment]
Another question, can an Iceberg REST server serve multiple cloud users?

File: core/src/main/java/org/apache/gravitino/credential/config/GCSCredentialConfig.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import javax.annotation.Nullable;
+import org.apache.gravitino.Config;
+import org.apache.gravitino.config.ConfigBuilder;
+import org.apache.gravitino.config.ConfigConstants;
+import org.apache.gravitino.config.ConfigEntry;
+import org.apache.gravitino.storage.GCSProperties;
+
+public class GCSCredentialConfig extends Config {
+  public static final ConfigEntry<String> GCS_CREDENTIAL_FILE_PATH =
+      new ConfigBuilder(GCSProperties.GRAVITINO_GCS_CREDENTIAL_FILE_PATH)
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/credential/config/GCSCredentialConfig.java:32]: it's more extensible, different catalogs could use different credentials.

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java (hunk 5 above)
Comment (FANNG1) [reply bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java:135]: ok

**Comment by FANNG1** [comment]
> Will the Gravitino fileset reuse this token provider? If so, credential config file isn't a good choice, because if we have different fileset catalog, they maybe have different different GCS credential files.

 Gravitino fileset will reuse this token provider, and different fileset catalogs could use different credential config file.

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java (hunk 4 above)
Comment (FANNG1) [reply bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java:75]: I prefer to return null, the upper layer could process this, WDYT?

**Comment by jerqi** [comment]
> > Another question, can an Iceberg REST server serve multiple cloud users?
> 
> You mean one cloud user corresponding to one metalake? Not support yet.

No. An Iceberg REST server is used by multiple clouds at the same time. It's not related to the metalake.

File: core/build.gradle.kts [hunk 7]
```diff
@@ -24,6 +24,7 @@ plugins {
 
 dependencies {
   implementation(project(":api"))
+  implementation(project(":catalogs:catalog-common"))
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply core/build.gradle.kts]: removed

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java (hunk 5 above)
Comment (FANNG1) [reply bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java:135]: added

**Comment by FANNG1** [comment]
> > > Will the Gravitino fileset reuse this token provider? If so, credential config file isn't a good choice, because if we have different fileset catalog, they maybe have different different GCS credential files.
> > 
> > 
> > Gravitino fileset will reuse this token provider, and different fileset catalogs could use different credential config file.
> 
> I can't get the point why we use the file. It's not convenient for users.

then what's your option?

**Comment by FANNG1** [comment]
> > > Another question, can an Iceberg REST server serve multiple cloud users?
> > 
> > 
> > You mean one cloud user corresponding to one metalake? Not support yet.
> 
> No. An Iceberg REST server is used by multiple clouds at the same time. It's not related to the metalake.

yes, different catalog could use different credential providers

**Comment by jerqi** [comment]
> > > > Will the Gravitino fileset reuse this token provider? If so, credential config file isn't a good choice, because if we have different fileset catalog, they maybe have different different GCS credential files.
> > > 
> > > 
> > > Gravitino fileset will reuse this token provider, and different fileset catalogs could use different credential config file.
> > 
> > 
> > I can't get the point why we use the file. It's not convenient for users.
> 
> then what's your option?

Could it be the properties of the Gravitino catalog？

**Comment by FANNG1** [comment]
@jerryshao @yuqi1129 @jerqi any other comments?

File: gradle/libs.versions.toml [hunk 8]
```diff
@@ -178,6 +179,7 @@ iceberg-core = { group = "org.apache.iceberg", name = "iceberg-core", version.re
 iceberg-api = { group = "org.apache.iceberg", name = "iceberg-api", version.ref = "iceberg" }
 iceberg-hive-metastore = { group = "org.apache.iceberg", name = "iceberg-hive-metastore", version.ref = "iceberg" }
 iceberg-gcp = { group = "org.apache.iceberg", name = "iceberg-gcp", version.ref = "iceberg" }
+iceberg-gcp-bundle = { group = "org.apache.iceberg", name = "iceberg-gcp-bundle", version.ref = "iceberg" }
```
Comment (yuqi1129) [inline gradle/libs.versions.toml:182]: Have you added this to the file `LICENSE.bin`?

Comment (yuqi1129) [reply gradle/libs.versions.toml:182]: ditto

File: iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuqi1129) [reply iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java:36]: To enable this test, we need to combine these three variants.?

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+          CredentialAccessBoundary.AccessBoundaryRule.Builder builder =
+              CredentialAccessBoundary.AccessBoundaryRule.newBuilder();
+          builder.setAvailableResource(toGCSBucketResource(bucket));
+          builder.setAvailabilityCondition(
+              CredentialAccessBoundary.AccessBoundaryRule.AvailabilityCondition.newBuilder()
+                  .setExpression(String.join(" || ", writeConditions))
+                  .build());
+          builder.setAvailablePermissions(Arrays.asList("inRole:roles/storage.legacyBucketWriter"));
+          credentialAccessBoundaryBuilder.addRule(builder.build());
+        });
```
Comment (yuqi1129) [inline bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java:180]: These two parts are almost the same, could you extract a common method?

File: iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/integration/test/IcebergRESTGCSIT.java:36]: yes

**Review by jerqi — APPROVED** [review]
LGTM.

File: gradle/libs.versions.toml (hunk 8 above)
Comment (FANNG1) [reply gradle/libs.versions.toml:182]: ok

File: bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java (hunk 9 above)
Comment (FANNG1) [reply bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java]: updated

File: gradle/libs.versions.toml (hunk 8 above)
Comment (FANNG1) [reply gradle/libs.versions.toml:182]: updated

**Comment by FANNG1** [comment]
@yuqi1129 please help to review again

**Review by yuqi1129 — APPROVED** [review]

**Review by yuqi1129 — APPROVED** [review]

**Pull request merged**